scylla-cdc-printer/src/main.rs

```diff
@@ -1,11 +1,73 @@
 use std::error::Error;
-use std::time::Duration;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use chrono::Duration;
+use core::time;
+use scylla::{Session, SessionBuilder};
+use tokio::time::sleep;
+
+use scylla_cdc::reader::Reader;
+use scylla_cdc::stream_generation::GenerationFetcher;
+
+const SLEEP_INTERVAL: u64 = 1;
+const WINDOW_SIZE: i64 = 2;
+const SAFETY_INTERVAL: i64 = 1;
+
+struct CDCLogPrinter {
+    session: Arc<Session>,
+    sleep_interval: time::Duration,
+    window_size: chrono::Duration,
+    safety_interval: chrono::Duration,
+}
+
+impl CDCLogPrinter {
+    fn new(session: &Arc<Session>) -> CDCLogPrinter {
+        CDCLogPrinter {
+            session: Arc::clone(session),
+            sleep_interval: time::Duration::new(SLEEP_INTERVAL, 0),
+            window_size: chrono::Duration::seconds(WINDOW_SIZE),
+            safety_interval: chrono::Duration::seconds(SAFETY_INTERVAL),
+        }
+    }
+
+    async fn run(&self, keyspace: String, table_name: String, start_timestamp: chrono::Duration) -> Result<(), Box<dyn Error>> {
+        let fetcher = GenerationFetcher::new(&self.session);
+        let mut generation = fetcher.fetch_generation_by_timestamp(&start_timestamp).await?.unwrap_or(loop {
+            match fetcher.fetch_all_generations().await?.last() {
+                None => { sleep(self.sleep_interval).await; }
+                Some(generation) => break generation.clone()
+            }
+        });
+        loop {
+            let stream_ids = fetcher.fetch_stream_ids(&generation).await?;
+            for stream_id in stream_ids {
+                let reader = Reader::new(vec![stream_id.clone()], start_timestamp, self.window_size, self.safety_interval, self.sleep_interval);
+                reader.fetch_cdc(keyspace.clone(), table_name.clone()).await?;
+            }
```

Comment:
This doesn't work well. Without the end timestamp set, `fetch_cdc` will keep reading its stream indefinitely. Moreover, what you are doing here is: run `fetch_cdc` for one stream, then after it finishes run it for another stream, etc. There is no concurrency here.

The big idea is that we want to run each reader in a task, independently from each other. This will allow them to utilize multiple cores. In the meantime, your main function monitors if there is a new generation on the horizon - if so, then you should set the end timestamp for all readers, wait until they finish reading and then proceed to the next generation.

Note that being able to concurrently run `fetch_cdc` in one task and calling `set_upper_timestamp` requires this suggestion to be applied to the PR with stream reader: https://github.com/piodul/scylla-cdc-rust/pull/9#discussion_r746949292